compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java

```diff
@@ -1161,11 +1168,23 @@ private BType getExpectedTypeForInferredTypedescMember(BUnionType originalType,
             return null;
         }
 
-        if (types.size() == 1) {
-            return types.iterator().next();
+        // Add the original union type if all the members of the original type are present in the `types` list.
+        LinkedHashSet<BType> originalTypesSet = new LinkedHashSet<>(); // This is to maintain the order of members
```

Comment:
I would use a different name. Since it isn't really/exactly the original types.

Reply:
Will create an issue to this